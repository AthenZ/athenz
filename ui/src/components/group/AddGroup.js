/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import Button from '../denali/Button';
import Input from '../denali/Input';
import InputLabel from '../denali/InputLabel';
import Member from '../member/Member';
import styled from '@emotion/styled';
import { colors } from '../denali/styles';
import AddModal from '../modal/AddModal';
import DateUtils from '../utils/DateUtils';
import RequestUtils from '../utils/RequestUtils';

const SectionDiv = styled.div`
    align-items: flex-start;
    display: flex;
    flex-flow: row nowrap;
    padding: 10px 10px;
`;

const StyledInputLabel = styled(InputLabel)`
    float: left;
    font-size: 14px;
    font-weight: 700;
    padding-top: 12px;
    width: 17%;
`;

const StyledInput = styled(Input)`
    width: 500px;
`;

const StyledInputUser = styled(Input)`
    margin-top: 5px;
`;

const ContentDiv = styled.div`
    flex: 1 1;
    margin-right: 10px;
`;

const AddMemberDiv = styled.div`
    display: flex;
`;

const StyledIncludedMembersDiv = styled.div`
    width: 65%;
`;

const SectionsDiv = styled.div`
    width: 780px;
    text-align: left;
    background-color: ${colors.white};
`;

const ButtonDiv = styled.div`
    margin-left: 10px;
`;

const StyledButton = styled(Button)`
    width: 125px;
`;

export default class AddGroup extends React.Component {
    constructor(props) {
        super(props);
        this.api = props.api;
        this.addMember = this.addMember.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.dateUtils = new DateUtils();
        this.state = {
            saving: 'nope',
            name: '',
            newMemberName: '',
            members: [],
            justification: '',
        };
    }

    inputChanged(key, evt) {
        let value = '';
        if (evt.target) {
            value = evt.target.value;
        } else {
            value = evt ? evt : '';
        }
        this.setState({ [key]: value });
    }

    getJustification() {
        if (this.props.justificationRequired) {
            return (
                <SectionDiv>
                    <StyledInputLabel>Justification</StyledInputLabel>
                    <ContentDiv>
                        <StyledInput
                            id='justification'
                            name='justification'
                            value={this.state.justification}
                            onChange={this.inputChanged.bind(
                                this,
                                'justification'
                            )}
                            autoComplete={'off'}
                            placeholder='Enter justification here'
                        />
                    </ContentDiv>
                </SectionDiv>
            );
        }
    }

    addMember() {
        let name = this.state.newMemberName;
        let members = this.state.members;

        if (!name) {
            return;
        }
        let names = (name || '')
            .replace(/[\r\n\s]+/g, ',')
            .split(',')
            .map((n) => n.trim())
            .filter((n) => n);

        for (let i = 0; i < names.length; i++) {
            members.push({
                memberName: names[i],
            });
        }

        this.setState({
            members,
            newMemberName: '',
        });
    }

    deleteMember(idx) {
        let members = this.state.members;
        // this if is done to avoid [null] condition
        if (members.length === 1) {
            members = null;
        } else {
            delete members[idx];
        }
        this.setState({ members });
    }

    onSubmit() {
        let groupName = this.state.name;
        if (!groupName || groupName === '') {
            this.setState({
                errorMessage: 'Group name is required.',
            });
            return;
        }

        let group = { name: groupName };
        group.groupMembers =
            this.state.members.filter((member) => {
                return member != null || member != undefined;
            }) || [];
        if (this.state.newMemberName && this.state.newMemberName !== '') {
            let names = (this.state.newMemberName || '')
                .replace(/[\r\n\s]+/g, ',')
                .split(',')
                .map((n) => n.trim())
                .filter((n) => n);

            for (let i = 0; i < names.length; i++) {
                group.groupMembers.push({
                    memberName: names[i],
                });
            }
        }

        if (
            this.props.justificationRequired &&
            (this.state.justification === undefined ||
                this.state.justification.trim() === '')
        ) {
            this.setState({
                errorMessage: 'Justification is required to add a group.',
            });
            return;
        }

        this.api
            .getGroups(this.props.domain)
            .then((groups) => {
                if (groups.includes(groupName)) {
                    this.setState({
                        errorMessage: 'Group already exists.',
                    });
                    return;
                }
                let auditRef = this.state.justification
                    ? this.state.justification
                    : ''; // no UX for this
                this.api
                    .addGroup(
                        this.props.domain,
                        groupName,
                        group,
                        auditRef,
                        this.props._csrf
                    )
                    .then(() => {
                        this.props.onSubmit(
                            `Successfully created group ${groupName}`
                        );
                    })
                    .catch((err) => {
                        let message = '';
                        if (err.statusCode === 0) {
                            message = 'Okta expired. Please refresh the page';
                        } else {
                            message = `Status: ${err.statusCode}. Message: ${err.body.message}`;
                        }
                        this.setState({
                            errorMessage: message,
                        });
                    });
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    render() {
        let memberNameChanged = this.inputChanged.bind(this, 'newMemberName');

        let nameChanged = this.inputChanged.bind(this, 'name');

        let members = this.state.members
            ? this.state.members.map((item, idx) => {
                  // dummy place holder so that it can be be used in the form
                  item.approved = true;
                  let remove = this.deleteMember.bind(this, idx);
                  return (
                      <Member
                          key={idx}
                          item={item}
                          onClickRemove={remove}
                          noanim
                      />
                  );
              })
            : '';
        let sections = (
            <SectionsDiv>
                <SectionDiv>
                    <StyledInputLabel>Group Name</StyledInputLabel>
                    <ContentDiv>
                        <StyledInput
                            placeholder='Enter New Group Name'
                            value={this.state.name}
                            onChange={nameChanged}
                            noanim
                            fluid
                        />
                    </ContentDiv>
                </SectionDiv>
                {
                    <SectionDiv>
                        <StyledInputLabel>Add Member(s)</StyledInputLabel>
                        <ContentDiv>
                            <AddMemberDiv>
                                <StyledInputUser
                                    placeholder='user.<userid> or <domain>.<service>'
                                    value={this.state.newMemberName}
                                    onChange={memberNameChanged}
                                    noanim
                                    fluid
                                />
                                <ButtonDiv>
                                    <StyledButton onClick={this.addMember}>
                                        Add
                                    </StyledButton>
                                </ButtonDiv>
                            </AddMemberDiv>
                        </ContentDiv>
                    </SectionDiv>
                }
                <SectionDiv>
                    <StyledInputLabel />
                    <StyledIncludedMembersDiv>
                        {members}
                    </StyledIncludedMembersDiv>
                </SectionDiv>
                {this.getJustification()}
            </SectionsDiv>
        );
        return (
            <div data-testid='add-group'>
                <AddModal
                    isOpen={this.props.showAddGroup}
                    cancel={this.props.onCancel}
                    submit={this.onSubmit}
                    title={`Add Group to ${this.props.domain}`}
                    errorMessage={this.state.errorMessage}
                    sections={sections}
                />
            </div>
        );
    }
}
