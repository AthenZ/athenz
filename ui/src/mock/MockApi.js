const origApi = require('../api').default;

export default class MockApi {
    static setMockApi(mockApi) {
        require('../api').default = () => {
            return mockApi;
        };
    }
    static cleanMockApi() {
        require('../api').default = () => {
            return origApi();
        };
    }
}
