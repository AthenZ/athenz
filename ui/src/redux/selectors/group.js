import { getFullName, mapToList } from '../utils';
import { groupDelimiter } from '../config';

export const thunkSelectGroups = (state) => {
    return state.groups.groups ? state.groups.groups : {};
};

export const thunkSelectGroup = (state, domainName, groupName) => {
    return selectGroup(state, domainName, groupName);
};

export const thunkSelectGroupMembers = (state, groupName) => {
    return state.groups.groups && state.groups.groups[groupName]
        ? state.groups.groups[groupName].groupMembers
        : {};
};

export const thunkSelectGroupHistory = (state, domainName, groupName) => {
    return selectGroupHistory(state, domainName, groupName);
};

export const thunkSelectGroupRoleMembers = (state) => {
    return selectGroupRoleMembers(state);
};

export const selectGroups = (state) => {
    return state.groups.groups ? mapToList(state.groups.groups) : [];
};

export const selectGroup = (state, domainName, groupName) => {
    return state.groups.groups &&
        state.groups.groups[getFullName(domainName, groupDelimiter, groupName)]
        ? state.groups.groups[
              getFullName(domainName, groupDelimiter, groupName)
          ]
        : {};
};

export const selectGroupMembers = (state, domainName, groupName) => {
    return state.groups.groups &&
        state.groups.groups[getFullName(domainName, groupDelimiter, groupName)]
        ? mapToList(
              state.groups.groups[
                  getFullName(domainName, groupDelimiter, groupName)
              ].groupMembers
          )
        : [];
};

export const selectGroupHistory = (state, domainName, groupName) => {
    return state.groups.groups &&
        state.groups.groups[
            getFullName(domainName, groupDelimiter, groupName)
        ] &&
        state.groups.groups[getFullName(domainName, groupDelimiter, groupName)]
            .auditLog
        ? state.groups.groups[
              getFullName(domainName, groupDelimiter, groupName)
          ].auditLog
        : [];
};

export const selectGroupRoleMembers = (state) => {
    return state.groups.roleMembers ? state.groups.roleMembers : {};
};

export const selectGroupTags = (state, domainName, groupName) => {
    return state.groups.groups &&
        state.groups.groups[
            getFullName(domainName, groupDelimiter, groupName)
        ] &&
        state.groups.groups[getFullName(domainName, groupDelimiter, groupName)]
            .tags
        ? state.groups.groups[
              getFullName(domainName, groupDelimiter, groupName)
          ].tags
        : {};
};
