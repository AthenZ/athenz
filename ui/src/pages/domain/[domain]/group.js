/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import Header from '../../../components/header/Header';
import UserDomains from '../../../components/domain/UserDomains';
import API from '../../../api';
import styled from '@emotion/styled';
import Head from 'next/head';
import DomainDetails from '../../../components/header/DomainDetails';
import RequestUtils from '../../../components/utils/RequestUtils';
import Tabs from '../../../components/header/Tabs';
import Error from '../../_error';
import DomainNameHeader from '../../../components/header/DomainNameHeader';
import { connect } from 'react-redux';
import { getDomainData } from '../../../redux/thunks/domain';
import { getGroups } from '../../../redux/thunks/groups';
import GroupList from '../../../components/group/GroupList';
import { selectIsLoading } from '../../../redux/selectors/loading';
import { selectDomainData } from '../../../redux/selectors/domainData';
import Alert from '../../../components/denali/Alert';
import createCache from '@emotion/cache';
import { CacheProvider } from '@emotion/react';
import { ReduxPageLoader } from '../../../components/denali/ReduxPageLoader';
import { getAllUsers } from '../../../redux/thunks/user';
import { selectAllUsers } from '../../../redux/selectors/user';

const AppContainerDiv = styled.div`
    align-items: stretch;
    flex-flow: row nowrap;
    height: 100%;
    display: flex;
    justify-content: flex-start;
`;

const MainContentDiv = styled.div`
    flex: 1 1 calc(100vh - 60px);
    overflow: hidden;
    font: 300 14px HelveticaNeue-Reg, Helvetica, Arial, sans-serif;
`;

const GroupsContainerDiv = styled.div`
    align-items: stretch;
    flex: 1 1;
    height: calc(100vh - 60px);
    overflow: auto;
    display: flex;
    flex-direction: column;
`;

const GroupsContentDiv = styled.div``;

const PageHeaderDiv = styled.div`
    background: linear-gradient(to top, #f2f2f2, #fff);
    padding: 20px 30px 0;
`;

export async function getServerSideProps(context) {
    let api = API(context.req);
    let reload = false;
    let notFound = false;
    let error = null;
    const domains = await Promise.all([api.getForm()]).catch((err) => {
        let response = RequestUtils.errorCheckHelper(err);
        reload = response.reload;
        error = response.error;
        return [{}];
    });
    return {
        props: {
            reload,
            notFound,
            error,
            domainName: context.query.domain,
            userName: context.req.session.shortId,
            _csrf: domains[0],
            nonce: context.req.headers.rid,
        },
    };
}

class GroupPage extends React.Component {
    constructor(props) {
        super(props);
        this.api = API();
        this.cache = createCache({
            key: 'athenz',
            nonce: this.props.nonce,
        });
        this.state = {
            errorMessage: '',
            showError: false,
        };
        this.showError = this.showError.bind(this);
    }

    componentDidMount() {
        const {
            getAllUsers,
            getDomainData,
            getGroupsList,
            domainName,
            userName,
        } = this.props;
        Promise.all([
            getAllUsers(),
            getDomainData(domainName, userName),
            getGroupsList(domainName),
        ]).catch((err) => {
            this.showError(RequestUtils.fetcherErrorCheckHelper(err));
        });
    }

    showError(errorMessage) {
        this.setState({
            showError: true,
            errorMessage: errorMessage,
        });
    }

    render() {
        const { reload, _csrf, domainData, isLoading, domainName } = this.props;
        if (reload) {
            window.location.reload();
            return <div />;
        }
        if (this.props.error) {
            return <Error err={this.props.error} />;
        }

        if (this.state.showError) {
            return (
                <Alert
                    isOpen={this.state.showError}
                    title={this.state.errorMessage}
                    onClose={() => {}}
                    type='danger'
                />
            );
        }

        return isLoading.includes('getDomainData') ? (
            <ReduxPageLoader message={'Loading domain data'} />
        ) : (
            <CacheProvider value={this.cache}>
                <div data-testid='group'>
                    <Head>
                        <title>Athenz</title>
                    </Head>
                    <Header showSearch={true} />
                    <MainContentDiv>
                        <AppContainerDiv>
                            <GroupsContainerDiv>
                                <GroupsContentDiv>
                                    <PageHeaderDiv>
                                        <DomainNameHeader
                                            domainName={domainName}
                                        />
                                        <DomainDetails
                                            api={this.api}
                                            _csrf={_csrf}
                                        />
                                        <Tabs
                                            domain={domainName}
                                            selectedName={'groups'}
                                        />
                                    </PageHeaderDiv>
                                    <GroupList
                                        domain={domainName}
                                        _csrf={_csrf}
                                    />
                                </GroupsContentDiv>
                            </GroupsContainerDiv>
                            <UserDomains domain={domainName} />
                        </AppContainerDiv>
                    </MainContentDiv>
                </div>
            </CacheProvider>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        ...props,
        domainData: selectDomainData(state),
        isLoading: selectIsLoading(state),
        userList: selectAllUsers(state),
    };
};

const mapDispatchToProps = (dispatch) => ({
    getAllUsers: () => dispatch(getAllUsers()),
    getDomainData: (domainName, userName) =>
        dispatch(getDomainData(domainName, userName)),
    getGroupsList: (domainName) => dispatch(getGroups(domainName)),
});

export default connect(mapStateToProps, mapDispatchToProps)(GroupPage);
