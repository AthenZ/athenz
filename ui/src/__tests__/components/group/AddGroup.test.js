/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import {
    cleanup,
    fireEvent,
    getByTestId,
    screen,
    waitFor,
} from '@testing-library/react';
import AddGroup from '../../../components/group/AddGroup';
import {
    ADD_GROUP_AUDIT_ENABLED_TOOLTIP,
    GROUP_MEMBER_PLACEHOLDER,
} from '../../../components/constants/constants';
import {
    buildDomainDataForState,
    buildGroupsForState,
    getStateWithDomainData,
    getStateWithDomainDataAndUserList,
    getStateWithGroups,
    getStateWithUserList,
    renderWithRedux,
} from '../../../tests_utils/ComponentsTestUtils';
import MockApi from '../../../mock/MockApi';
import { singleApiGroup } from '../../redux/config/group.test';
import { act } from 'react-dom/test-utils';

describe('AddGroup', () => {
    afterEach(() => {
        MockApi.cleanMockApi();
        cleanup();
    });

    it('should render', () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const { getByTestId } = renderWithRedux(
            <AddGroup domain={domain} showAddGroup={true} onCancel={onClose} />,
            getStateWithDomainData(domainData)
        );
        const addGroupForm = getByTestId('add-group');
        expect(addGroupForm).toMatchSnapshot();
    });

    it('should throw error for group name 1', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainData(domainData)
        );
        expect(
            screen.getByPlaceholderText('Enter New Group Name')
        ).toBeInTheDocument();
        fireEvent.change(screen.getByPlaceholderText('Enter New Group Name'), {
            target: { value: 'testgroup:test' },
        });
        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );
        const button = getByText('Submit');
        fireEvent.click(button);
        await waitFor(() => {
            expect(
                screen.getByText("Group name doesn't match regex:", {
                    exact: false,
                })
            ).toBeInTheDocument();
        });
    });

    it('should throw error for group name 2', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainData(domainData)
        );
        expect(
            screen.getByPlaceholderText('Enter New Group Name')
        ).toBeInTheDocument();

        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );
        fireEvent.change(screen.getByPlaceholderText('Enter New Group Name'), {
            target: { value: 'group(123)' },
        });
        const button = getByText('Submit');
        fireEvent.click(button);
        await waitFor(() => {
            expect(
                screen.getByText("Group name doesn't match regex:", {
                    exact: false,
                })
            ).toBeInTheDocument();
        });
    });

    it('should throw error for group name 3', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainData(domainData)
        );
        expect(
            screen.getByPlaceholderText('Enter New Group Name')
        ).toBeInTheDocument();

        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );
        fireEvent.change(screen.getByPlaceholderText('Enter New Group Name'), {
            target: { value: 'group/123' },
        });
        const button = getByText('Submit');
        fireEvent.click(button);
        await waitFor(() => {
            expect(
                screen.getByText("Group name doesn't match regex:", {
                    exact: false,
                })
            ).toBeInTheDocument();
        });
    });

    it('should not throw error for group name', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';
        const api = {
            addGroup: jest.fn(() => Promise.resolve(singleApiGroup)),
        };
        MockApi.setMockApi(api);
        const groups = buildGroupsForState({}, domain);
        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
                groupNames={'group1'}
            />,
            getStateWithGroups(groups, { domainData: domainData })
        );
        expect(
            screen.getByPlaceholderText('Enter New Group Name')
        ).toBeInTheDocument();

        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );
        const button = getByText('Submit');
        fireEvent.change(screen.getByPlaceholderText('Enter New Group Name'), {
            target: { value: 'group.test' },
        });
        fireEvent.click(button);
        await waitFor(() => {
            expect(onSubmit.mock.calls.length).toBe(1);
        });
    });

    it('should throw error for group member name 1', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const userList = { userList: [] };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainDataAndUserList(domainData, userList)
        );
        expect(
            screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER)
        ).toBeInTheDocument();

        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );

        await act(async () => {
            fireEvent.change(
                screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER),
                {
                    target: { value: 'home.test:group.test' },
                }
            );
        });

        await waitFor(() => {
            expect(
                screen.getByText('home.test:group.test')
            ).toBeInTheDocument();
        });
        fireEvent.click(screen.getByText('home.test:group.test'));

        const button = getByText('Add');
        fireEvent.click(button);
        await waitFor(() => {
            expect(
                screen.getByText("Member name doesn't match regex: ", {
                    exact: false,
                })
            ).toBeInTheDocument();
        });
    });

    it('should throw error for group member name 2', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const userList = { userList: [] };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainDataAndUserList(domainData, userList)
        );
        expect(
            screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER)
        ).toBeInTheDocument();

        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );

        await act(async () => {
            fireEvent.change(
                screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER),
                {
                    target: { value: 'user.test1.' },
                }
            );
        });

        await waitFor(() => {
            expect(screen.getByText('user.test1.')).toBeInTheDocument();
        });
        fireEvent.click(screen.getByText('user.test1.'));

        const button = getByText('Add');
        fireEvent.click(button);
        await waitFor(() => {
            expect(
                screen.getByText("Member name doesn't match regex: ", {
                    exact: false,
                })
            ).toBeInTheDocument();
        });
    });

    it('should throw error for group member name 3', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const userList = { userList: [] };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainDataAndUserList(domainData, userList)
        );
        expect(
            screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER)
        ).toBeInTheDocument();

        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );

        await act(async () => {
            fireEvent.change(
                screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER),
                {
                    target: { value: 'unix.service.test/' },
                }
            );
        });

        await waitFor(() => {
            expect(screen.getByText('unix.service.test/')).toBeInTheDocument();
        });
        fireEvent.click(screen.getByText('unix.service.test/'));

        const button = getByText('Add');
        fireEvent.click(button);
        await waitFor(() => {
            expect(
                screen.getByText("Member name doesn't match regex: ", {
                    exact: false,
                })
            ).toBeInTheDocument();
        });
    });

    it('should not throw error for group member name', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';
        const api = {
            addGroup: jest.fn(() => Promise.resolve({ name: 'group.test' })),
        };
        const groups = buildGroupsForState({}, domain);
        MockApi.setMockApi(api);
        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithGroups(groups, { domainData: domainData })
        );
        expect(
            screen.getByPlaceholderText('Enter New Group Name')
        ).toBeInTheDocument();

        fireEvent.change(
            screen.getByPlaceholderText('Enter justification here'),
            {
                target: { value: 'justification' },
            }
        );
        const button = getByText('Submit');
        fireEvent.change(screen.getByPlaceholderText('Enter New Group Name'), {
            target: { value: 'group.test' },
        });
        fireEvent.change(
            screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER),
            {
                target: { value: 'user.test' },
            }
        );
        fireEvent.click(button);
        await waitFor(() => {
            expect(onSubmit.mock.calls.length).toBe(1);
        });
    });

    it('should have auditEnabled switch', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainData(domainData)
        );

        expect(
            screen.getByTestId('auditEnabled-switch-input')
        ).toBeInTheDocument();

        expect(screen.getAllByTestId('auditEnabled-switch-input')).toHaveLength(
            1
        );

        fireEvent.click(screen.getByTestId('auditEnabled-switch-input'));

        await waitFor(() => {
            expect(
                screen.getAllByTitle(ADD_GROUP_AUDIT_ENABLED_TOOLTIP)
            ).toHaveLength(2);
        });
    });

    it('should have auditEnabled switch disabled after user is added', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const userList = { userList: [{ login: 'test', name: 'Test User' }] };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByText } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainDataAndUserList(domainData, userList)
        );

        await act(async () => {
            fireEvent.change(
                screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER),
                {
                    target: { value: 'user.test' },
                }
            );
        });

        await waitFor(() => {
            expect(
                screen.getByText('Test User [user.test]')
            ).toBeInTheDocument();
        });
        fireEvent.click(screen.getByText('Test User [user.test]'));

        const button = getByText('Add');
        fireEvent.click(button);

        await waitFor(() => {
            expect(
                screen.getByTestId('auditEnabled-switch-input')
            ).toBeDisabled();
        });
    });

    it('should open modal with auditEnabled switch', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: true,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByTestId } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainData(domainData)
        );

        await waitFor(() => {
            expect(
                screen.queryByTestId('auditEnabled-switch-input')
            ).toBeEnabled();
        });
    });

    it('should open modal without auditEnabled switch', async () => {
        let domain = 'domain';
        const domainMetadata = {
            auditEnabled: false,
        };
        const domainData = buildDomainDataForState(domainMetadata, domain);
        const onClose = jest.fn();
        const onSubmit = jest.fn();
        const _csrf = '_csrf';

        const { getByTestId } = renderWithRedux(
            <AddGroup
                _csrf={_csrf}
                onSubmit={onSubmit}
                domain={domain}
                showAddGroup={true}
                onCancel={onClose}
            />,
            getStateWithDomainData(domainData)
        );

        const auditEnabledSwitch = screen.queryByText('Audit Enabled');
        expect(auditEnabledSwitch).toBeNull();
    });

    it('search member add group', async () => {
        let domain = 'domain';
        let userList = { userList: [{ login: 'mock', name: 'Mock User' }] };
        const onCancelMock = jest.fn();
        renderWithRedux(
            <AddGroup
                domain={domain}
                showAddGroup={true}
                onCancel={onCancelMock}
            />,
            getStateWithUserList(userList)
        );

        // change input to mocked user
        await act(async () => {
            fireEvent.change(
                screen.getByPlaceholderText(GROUP_MEMBER_PLACEHOLDER),
                {
                    target: { value: 'mock' },
                }
            );
        });

        // verify the correct input 'Mock User [user.mock]' is presented
        await waitFor(() =>
            expect(
                screen.getByText('Mock User [user.mock]')
            ).toBeInTheDocument()
        );
    });
});
