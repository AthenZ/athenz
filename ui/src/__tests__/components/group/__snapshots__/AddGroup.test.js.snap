// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddGroup should render 1`] = `
<div
  data-testid="add-group"
/>
`;
